File: Source/Engine/Graphics/SwapChain.cpp
// Copyright © Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

#include "Graphics/SwapChain.h"
#include "Graphics/Texture.h"
#include "Graphics/Graphics.h"
#include "Core/Assert.h"

namespace Alimer
{
	SwapChain::SwapChain(const SwapChainCreateInfo& info)
		: width(info.width)
		, height(info.height)
		, colorFormat(info.colorFormat)
		, presentMode(info.presentMode)
	{
	}

    SwapChainRef SwapChain::Create(void* window, const SwapChainCreateInfo& info)
    {
        ALIMER_ASSERT(gGraphics().IsInitialized());
        ALIMER_ASSERT(window != nullptr);

        return gGraphics().CreateSwapChain(window, info);
    }

	void SwapChain::Resize(uint32_t width_, uint32_t height_)
	{
		width = width_;
		height = height_;

		ResizeBackBuffer(width, height);
	}
}

File: src/Alimer/Graphics/SwapChain.cs
// Copyright (c) Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

namespace Alimer.Graphics;

public abstract class SwapChain : GraphicsObject
{
    public SwapChain(in SwapChainDescriptor descriptor)
        : base(descriptor.Label)
    {
        Surface = descriptor.Surface;
        Width = descriptor.Width;
        Height = descriptor.Height;
        ColorFormat = descriptor.Format;
        PresentMode = descriptor.PresentMode;
    }

    public SwapChainSurface Surface { get; }

    public uint Width { get; private set; }
    public uint Height { get; private set; }
    public PixelFormat ColorFormat { get; protected set; }
    public PresentMode PresentMode { get; }

    public void Resize(uint width, uint height)
    {
        if (Width == width && Height == height)
            return;

        ResizeBackBuffer();
        Width = width;
        Height = height;
    }

    protected abstract void ResizeBackBuffer();

    /// <summary>
    /// Acquires the next available texture for rendering or processing operations.
    /// </summary>
    /// <returns>A <see cref="Texture"/> object representing the next available texture, or <see langword="null"/> if no texture
    /// is currently available.</returns>
    public abstract Texture? AcquireNextTexture();
}
